### ISSUE TITLE
Restructure build matrix

### ISSUE THREAD

**Issue by nipafx** [junit-pioneer/junit-pioneer#474]
Our current build configuration operates on the following sets:

* operating systems `OS` = { Ubuntu, Windows, macOs }
* Java versions `J` = { 8, 11, 16 }
* modularity `M` = { `true`, `false` }
* JUnit versions `U` = { 5.7.0, 5.7.1, 5.8.0-M1 }

But we don't build all possible combinations `OS x J x M x U`, which would be 54 builds and slightly above the [20 builds GitHub runs in parallel for free](https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits). Even if we upgrade to team, which raises the limit to 60 (and costs 4$/member/month), we'd be just one additional Java or JUnit version away from going overboard.

Instead we build:

* `OS x J x { false } x { 5.7.1 }` in `main-build.yml` (9 builds)
* `OS x { 11, 16 } x { true } x { 5.7.1 }` in `module-build.yml` (6 builds)
* `{ Ubuntu } x { 8 } x { false } x U` in `junit5-build.yml` (3 builds)

I find that combination of builds somewhat arbitrary. Also, just one more Java or JUnit version and we blow the limit of 20 builds and need to wait twice as long (worst case) for the entire set to finish.

Can we find a better way to create builds?

**Comment by nipafx** [comment]
An obvious solution would be to define a baseline combination and then explore each set from there, keeping the others fixed, i.e.:

* `{ Ubuntu } x { 8 } x { false } x { 5.7.1 }` as baseline (1 build)
* `OS x { 8 } x { false } x { 5.7.1 }` (`|OS|` builds)
* `{ Ubuntu } x J x { false } x { 5.7.1 }` (`|J|` builds)
* `{ Ubuntu } x { 8 } x M x { 5.7.1 }` (2 builds)
* `{ Ubuntu } x { 8 } x { false } x U` (`|U|` builds)

This would result in `3 + |OS| + |J| + |U|` builds, which means it's not gonna tip over 20 any time soon. It's also structurally pleasing (very important!).

But does it miss important combinations? One could argue that testing on all `OS x J` is kinda important...

**Comment by aepfli** [comment]
I would also evaluate what do we really need on a merge request and what can be handled as an action on a merge into develop.

Generally speaking it is nice, if everything runs during a merge request - but is it really needed or is it needed all the time?

For verification if the code works (which imho the CI should guarantee) a small build is enough (to be honest we need it to get reminded of spotless).

What if we define two sets of CI jobs - the ones which run all the time, short fast, small set - and the ones we need to run before merging. It will not block a contributor and the contributor will get fast feedback. On the other hand, we as maintainer could eg. label or comment on the MR that we are `ready to merge` which could kickoff a bunch of more tests. Those builds are normally not timely critical, therefore we could even have more builds and exceed the limit.

This is just a suggestion and some thoughts about this. eg. GitLab is doing this for their Repositories too.

**Comment by aepfli** [comment]
another thought which came to my mind, we could merge the JAVA VERSION builds into one build with the tool chain support of gradle, and execute them sequentially.

Normally we would only have a little overhead and the test execution time, as gradle should be able to use the cached output of the previous gradle runs. At least worth an experiment to check feasibility

**Comment by Bukama** [comment]
We just decided to have a two tier build by default with `Ubunto`, `Java 8` and multiple `JUnit` versions by default and a full matrix build triggerable manually. We see value in the second tier (JUnit) because it's our main dependency and changes there have a high(er) chance to break something internal (e..g test kit) and external (extensions).

If possible we want to trigger builds by labels, so if someone opens a big change draft and wants to run full matrix he should be able to do this easily.

**Comment by aepfli** [comment]
```
if: contains(github.event.pull_request.labels.*.name, 'some-defined-label')
```
https://github.community/t/do-something-if-a-particular-label-is-set/17149/2

As a note, might be worth to investigate into, and seems like an easy and feasible solution

**Comment by aepfli** [comment]
something like 
![image](https://user-images.githubusercontent.com/9987394/120068504-4e2fc600-c081-11eb-8237-51f7a440332c.png)

with an own action which fails, when the maintainer run is missing?

**Comment by aepfli** [comment]
here is a little sneak peak:
https://github.com/aepfli/junit-pioneer/pull/9

if the mr does not have the label merge-ready, it will trigger a minified build. with a failing action notifying us, that a maintainer needs to do something.
![image](https://user-images.githubusercontent.com/9987394/120076307-310cee80-c0a5-11eb-8c11-11e67090d4d0.png)


if it has the label the whole pipeline is executed.
![image](https://user-images.githubusercontent.com/9987394/120076293-26525980-c0a5-11eb-8883-6c3d31124280.png)

The good part is, that now everything is happening in one pipeline. it is still reliable fast for MRs for our users, but offers us the full execution of all scenarios we want

**Comment by aepfli** [comment]
@nipafx @Bukama @beatngu13 what do you think about such an approach? good bad ugly?

**Comment by aepfli** [comment]
so, i was bored, and added a finalizing step with Java 17 ea and gradle RC - also on windows, linux and macos :D 

![image](https://user-images.githubusercontent.com/9987394/120228039-a2d26d00-c24a-11eb-9c89-e5d676fbb43e.png)

**Comment by beatngu13** [comment]
>What if we define two sets of CI jobs - the ones which run all the time, short fast, small set - and the ones we need to run before merging. It will not block a contributor and the contributor will get fast feedback.

I like that idea!

When it comes to the [current implementation](https://github.com/aepfli/junit-pioneer/pull/9), the workflow file looks a bit "packed" at first sight. (The duplicated caching stuff causes quite some noise… 😕)

I think I tend to prefer multiple workflows over a single workflow with multiple jobs, but I'm not sure if that is feasible and/or idiomatic (related: [Single workflow vs multiple workflows](https://github.community/t/single-workflow-vs-multiple-workflows/136516)).

**Comment by Bukama** [comment]
I like!

**Comment by aepfli** [comment]
https://github.com/junit-pioneer/junit-pioneer/actions/runs/897069483 is an action as an example and #495 is the pull request
